File: exitcode.go
package agentsdk

import "fmt"

// Semantic exit codes. Each code maps to a distinct failure category so that
// calling agents (shells, orchestrators) can branch on the numeric exit code.
const (
	ExitSuccess       = 0 // 正常完成
	ExitFatalError    = 1 // 未分类的致命错误
	ExitInvalidParams = 2 // 输入参数无效
	ExitNotFound      = 3 // 资源未找到
	ExitNetworkError  = 4 // 网络错误
	ExitLockConflict  = 5 // 资源互斥锁定
)

// ExitError wraps an error with a semantic exit code.
// It implements the error interface so it can be returned from any function
// that returns error, while carrying the exit code for os.Exit.
type ExitError struct {
	Code int
	Err  error
}

// Error returns "exit N: message" format for logging and display.
func (e *ExitError) Error() string {
	return fmt.Sprintf("exit %d: %s", e.Code, e.Err)
}

// Unwrap returns the wrapped error for errors.Is/As chaining.
func (e *ExitError) Unwrap() error {
	return e.Err
}

// errorCodeEntry is a registry entry mapping an error_code string to an
// exit code and human-readable description.
type errorCodeEntry struct {
	ExitCode    int
	Description string
}

// builtInErrorCodes are the default error codes that ship with the SDK.
// These cannot be overridden via Register.
var builtInErrorCodes = map[string]errorCodeEntry{
	"FATAL_CRASH":    {ExitCode: ExitFatalError, Description: "程序崩溃（panic 恢复）"},
	"INTERNAL_ERROR": {ExitCode: ExitFatalError, Description: "内部错误"},
	"INPUT_INVALID":  {ExitCode: ExitInvalidParams, Description: "输入参数无效"},
	"NOT_FOUND":      {ExitCode: ExitNotFound, Description: "资源未找到"},
	"RESOURCE_LOCKED": {ExitCode: ExitLockConflict, Description: "资源互斥锁定"},
}

// ErrorCodeRegistry maps error_code strings (used in JSONL envelopes) to
// semantic exit codes. Built-in codes are protected from override.
type ErrorCodeRegistry struct {
	codes map[string]errorCodeEntry
}

// NewErrorCodeRegistry creates a registry pre-loaded with built-in error codes.
func NewErrorCodeRegistry() *ErrorCodeRegistry {
	r := &ErrorCodeRegistry{
		codes: make(map[string]errorCodeEntry, len(builtInErrorCodes)),
	}
	for k, v := range builtInErrorCodes {
		r.codes[k] = v
	}
	return r
}

// Register adds a custom error code to the registry.
// Returns an error if the code conflicts with a built-in code.
func (r *ErrorCodeRegistry) Register(code string, exitCode int, description string) error {
	if _, isBuiltIn := builtInErrorCodes[code]; isBuiltIn {
		return fmt.Errorf("cannot override built-in error code: %s", code)
	}
	r.codes[code] = errorCodeEntry{ExitCode: exitCode, Description: description}
	return nil
}

// Lookup returns the exit code and description for the given error_code.
// ok is false if the code is not registered.
func (r *ErrorCodeRegistry) Lookup(code string) (exitCode int, description string, ok bool) {
	entry, ok := r.codes[code]
	if !ok {
		return 0, "", false
	}
	return entry.ExitCode, entry.Description, true
}

// ToExitCode returns the exit code for the given error_code string.
// Falls back to ExitFatalError (1) if the code is not found.
func (r *ErrorCodeRegistry) ToExitCode(code string) int {
	if entry, ok := r.codes[code]; ok {
		return entry.ExitCode
	}
	return ExitFatalError
}

// AllCodes returns a copy of all registered codes (built-in + custom).
// Modifications to the returned map do not affect the registry.
func (r *ErrorCodeRegistry) AllCodes() map[string]errorCodeEntry {
	out := make(map[string]errorCodeEntry, len(r.codes))
	for k, v := range r.codes {
		out[k] = v
	}
	return out
}

File: sdks/python/agentsdk/exitcode.py
"""Exit code constants, ExitError exception, and ErrorCodeRegistry.

Provides the shared exit code contract between the SDK and consumers.
Built-in error codes are protected from override; custom codes may be
registered and re-registered (silently updates).
"""

# ---------------------------------------------------------------------------
# Exit code constants (must match Go SDK values 0-5)
# ---------------------------------------------------------------------------

EXIT_SUCCESS = 0
EXIT_FATAL_ERROR = 1
EXIT_INVALID_PARAMS = 2
EXIT_NOT_FOUND = 3
EXIT_NETWORK_ERROR = 4
EXIT_LOCK_CONFLICT = 5

# ---------------------------------------------------------------------------
# Built-in error codes registered by every ErrorCodeRegistry instance
# ---------------------------------------------------------------------------

_BUILTIN_CODES: dict[str, tuple[int, str]] = {
    "FATAL_CRASH": (EXIT_FATAL_ERROR, "Unrecoverable process crash"),
    "INTERNAL_ERROR": (EXIT_FATAL_ERROR, "Internal SDK error"),
    "INPUT_INVALID": (EXIT_INVALID_PARAMS, "Invalid input parameter"),
    "NOT_FOUND": (EXIT_NOT_FOUND, "Requested resource not found"),
    "RESOURCE_LOCKED": (EXIT_LOCK_CONFLICT, "Resource is locked by another process"),
}


class ExitError(Exception):
    """Exception carrying a semantic exit code and an optional wrapped error."""

    def __init__(self, code: int, message: str, original_error: Exception | None = None):
        self.code = code
        self.message = message
        self.original_error = original_error
        super().__init__(str(self))

    def __str__(self) -> str:
        return f"exit {self.code}: {self.message}"


class ErrorCodeRegistry:
    """Registry mapping string error codes to (exit_code, description) pairs.

    Five built-in codes are always present and **cannot** be overridden.
    Custom codes may be registered or re-registered (update semantics).
    """

    def __init__(self) -> None:
        # internal: code -> (exit_code, description)
        self._codes: dict[str, tuple[int, str]] = dict(_BUILTIN_CODES)

    # -- mutators -----------------------------------------------------------

    def register(self, code: str, exit_code: int, description: str) -> None:
        """Register a custom error code.  Raises ValueError for built-in codes."""
        if code in _BUILTIN_CODES:
            raise ValueError(f"Cannot override built-in error code: {code}")
        self._codes[code] = (exit_code, description)

    # -- accessors ----------------------------------------------------------

    def lookup(self, code: str) -> tuple[int, str, bool]:
        """Look up *code*.

        Returns ``(exit_code, description, found)``.
        Unknown codes yield ``(EXIT_FATAL_ERROR, "", False)``.
        """
        if code in self._codes:
            exit_code, description = self._codes[code]
            return (exit_code, description, True)
        return (EXIT_FATAL_ERROR, "", False)

    def to_exit_code(self, code: str) -> int:
        """Return the numeric exit code for *code*, or EXIT_FATAL_ERROR."""
        exit_code, _, found = self.lookup(code)
        if not found:
            return EXIT_FATAL_ERROR
        return exit_code

    def has_error_code(self, code: str) -> bool:
        """Return ``True`` if *code* is registered (built-in or custom)."""
        return self.lookup(code)[2]

    def all_codes(self) -> dict[str, tuple[int, str]]:
        """Return a **defensive copy** of all registered codes."""
        return dict(self._codes)
